launcher_go/v2/docker/commands.go

```diff
@@ -52,3 +57,197 @@ func (r *DockerBuilder) Run() error {
 	}
 	return nil
 }
+
+type DockerRunner struct {
+	Config      *config.Config
+	Ctx         *context.Context
+	ExtraEnv    []string
+	ExtraFlags  []string
+	Rm          bool
+	ContainerId string
+	CustomImage string
+	Cmd         []string
+	Stdin       io.Reader
+	SkipPorts   bool
+	DryRun      bool
+	Restart     bool
+	Detatch     bool
+	Hostname    string
+}
+
+func (r *DockerRunner) Run() error {
+	cmd := exec.CommandContext(*r.Ctx, utils.DockerPath, "run")
+
+	// Detatch signifies we do not want to supervise
+	if !r.Detatch {
+		cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+		cmd.Cancel = func() error {
+			if runtime.GOOS == "darwin" {
+				runCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+				stopCmd := exec.CommandContext(runCtx, utils.DockerPath, "stop", r.ContainerId)
+				utils.CmdRunner(stopCmd).Run()
+				cancel()
+			}
+			return unix.Kill(-cmd.Process.Pid, unix.SIGINT)
+		}
+	}
+	cmd.Env = r.Config.EnvArray(true)
+
+	if r.DryRun {
+		// multi-line env doesn't work super great from CLI, but we can print out the rest.
+		for k, v := range r.Config.Env {
+			if !strings.Contains(v, "\n") {
+				cmd.Args = append(cmd.Args, "--env")
+				cmd.Args = append(cmd.Args, k+"="+shellwords.Escape(v))
+			}
+		}
+	} else {
+		for k, _ := range r.Config.Env {
+			cmd.Args = append(cmd.Args, "--env")
+			cmd.Args = append(cmd.Args, k)
+		}
+	}
+
+	// Order is important here, we add extra env after config's env to override anything set in env.
+	for _, e := range r.ExtraEnv {
+		cmd.Args = append(cmd.Args, "--env")
+		cmd.Args = append(cmd.Args, e)
+	}
+	for k, v := range r.Config.Labels {
+		cmd.Args = append(cmd.Args, "--label")
+		cmd.Args = append(cmd.Args, k+"="+v)
+	}
+	if !r.SkipPorts {
+		for _, v := range r.Config.Expose {
+			if strings.Contains(v, ":") {
+				cmd.Args = append(cmd.Args, "-p")
+				cmd.Args = append(cmd.Args, v)
+			} else {
+				cmd.Args = append(cmd.Args, "--expose")
+				cmd.Args = append(cmd.Args, v)
+			}
+		}
+	}
+	for _, v := range r.Config.Volumes {
+		cmd.Args = append(cmd.Args, "-v")
+		cmd.Args = append(cmd.Args, v.Volume.Host+":"+v.Volume.Guest)
```

Comment:
```suggestion
		cmd.Args = append(cmd.Args, "--volume=" + v.Volume.Host + ":" + v.Volume.Guest)
```

Improved readability by using `--volume` instead of `-v`